pkg/operators/prometheus/prometheus.go

```diff
@@ -0,0 +1,163 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.opentelemetry.io/otel/exporters/prometheus"
+	"go.opentelemetry.io/otel/metric"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+type SetMetricsExporter interface {
+	SetMetricsExporter(metric.Meter)
+}
+
+const (
+	ParamEnableMetrics = "enable-metrics"
+	ParamMetricsID     = "metrics-id"
+	ParamListenAddress = "metrics-listen-address"
+	ParamMetricsPath   = "metrics-path"
+)
+
+type Prometheus struct {
+	exporter      *prometheus.Exporter
+	meterProvider metric.MeterProvider
+}
+
+func (l *Prometheus) EnrichEvent(a any) error {
+	return nil
+}
+
+func (l *Prometheus) Name() string {
+	return "Prometheus"
+}
+
+func (l *Prometheus) Description() string {
+	return "Provides a facility to export metrics using Prometheus"
+}
+
+func (l *Prometheus) Dependencies() []string {
+	return nil
+}
+
+func (l *Prometheus) GlobalParamDescs() params.ParamDescs {
+	return params.ParamDescs{
+		{
+			Key:          ParamListenAddress,
+			DefaultValue: "127.0.0.1:2223",
+			Description:  "Address to serve prometheus metrics on",
+		},
+		{
+			Key:          ParamMetricsPath,
+			DefaultValue: "/metrics",
+			Description:  "Path to export prometheus metrics on",
+		},
+	}
+}
+
+func (l *Prometheus) ParamDescs() params.ParamDescs {
+	return params.ParamDescs{
+		{
+			Key:          ParamEnableMetrics,
+			Title:        "Enable Stats Export",
+			DefaultValue: "false",
+			Description:  "Enables collecting stats from the gadget and export it via Prometheus",
+			IsMandatory:  true,
+			TypeHint:     params.TypeBool,
+		},
+		{
+			Key:         ParamMetricsID,
+			Title:       "Stats Identifier",
+			Description: "Will be used as part of the scope name for the metrics",
+		},
+	}
+}
+
+func (l *Prometheus) Init(globalParams *params.Params) error {
+	exporter, err := prometheus.New()
+	if err != nil {
+		return fmt.Errorf("initialize prometheus exporter: %w", err)
+	}
+	l.exporter = exporter
+	l.meterProvider = sdkmetric.NewMeterProvider(sdkmetric.WithReader(exporter))
+
+	listenAddress := globalParams.Get(ParamListenAddress).AsString()
+	metricsPath := globalParams.Get(ParamMetricsPath).AsString()
+
+	go func() {
+		mux := http.NewServeMux()
+		mux.Handle(metricsPath, promhttp.Handler())
+		err := http.ListenAndServe(listenAddress, mux)
+		if err != nil {
+			fmt.Printf("error serving http: %v", err)
+			return
+		}
+	}()
+	return nil
+}
+
+func (l *Prometheus) CanOperateOn(gadget gadgets.GadgetDesc) bool {
+	inst, ok := gadget.(gadgets.GadgetInstantiate)
+	if !ok {
+		return false
+	}
+	tempInstance, err := inst.NewInstance()
+	if err != nil {
```

Comment:
Hmm, good question. Some background: Operators have to do this call in order to check for instances on the gadget. The call should be lightweight as it could be used by each operator for each gadget. The only reason for `NewInstance()` (which basically should just return an uninitialized &Gadget{} pointer) to fail is when a gadget doesn't want to run in a certain environment (and we have no such case right now afaik). And if that would be the case, the error would be reported once you would want to run the gadget itself (which also leads to the NewInstance() call that is properly handled). So I think: the operator shouldn't worry about it - otherwise we'd have to do it in every operator, which would in most cases result in more than one warning per gadget.